newton/_src/sim/model.py

```diff
@@ -26,6 +30,123 @@
 from .state import State
 
 
+class ModelAttributeAssignment(IntEnum):
+    """Enumeration of attribute assignment categories.
+
+    Defines which component of the simulation system owns and manages specific attributes.
+    This categorization determines where custom attributes are attached during simulation
+    object creation (State, Control, or Contacts).
+    """
+
+    MODEL = 0
+    """Model attributes are attached to the Model object."""
+    STATE = 1
+    """State attributes are attached to the State object."""
+    CONTROL = 2
+    """Control attributes are attached to the Control object."""
+    CONTACT = 3
+    """Contact attributes are attached to the Contacts object."""
+
+
+class ModelAttributeFrequency(IntEnum):
+    """Enumeration of attribute frequency categories.
+
+    Defines the dimensional structure and indexing pattern for custom attributes.
+    This determines how many elements an attribute array should have and how it
+    should be indexed in relation to the model's entities such as joints, bodies, shapes, etc.
+    """
+
+    JOINT = 0
+    """Attribute frequency follows the number of joints (see :attr:`~newton.Model.joint_count`)."""
+    JOINT_DOF = 1
+    """Attribute frequency follows the number of joint degrees of freedom (see :attr:`~newton.Model.joint_dof_count`)."""
+    JOINT_COORD = 2
+    """Attribute frequency follows the number of joint positional coordinates (see :attr:`~newton.Model.joint_coord_count`)."""
+    BODY = 3
+    """Attribute frequency follows the number of bodies (see :attr:`~newton.Model.body_count`)."""
+    SHAPE = 4
+    """Attribute frequency follows the number of shapes (see :attr:`~newton.Model.shape_count`)."""
```

Comment:
@Milad-Rakhsha-NV I think we might need another type of frequency for attributes that are created only once, e.g. to track a basic scalar or vector at every state (mouse position, for example). So I recommend adding `ONCE` to this enum.